sources/lib/api/gcp/bigquery/_schema.py

```diff
@@ -136,53 +136,86 @@ def from_dataframe(dataframe, default_type='STRING'):
-    of dictionaries (in which case the first item is used), or a list of lists. In the latter
-    case the type of the elements is used and the field names are simply 'Column1', 'Column2', etc.
+    Infer a BigQuery table schema from a dictionary. If the dictionary has entries that
+    are in turn OrderedDicts these will be turned into RECORD types. Ideally this will
+    be an OrderedDict but it is not required.
 
     Args:
-      data: The list of dictionaries or lists.
+      data: The dict to infer a schema from.
     Returns:
       A list of dictionaries containing field 'name' and 'type' entries, suitable for use in a
           BigQuery Tables resource schema.
     """
-    if not data:
-      return []
-
-    def _get_type(value):
-      if isinstance(value, datetime.datetime):
-        return 'TIMESTAMP'
-      elif isinstance(value, bool):
-        return 'BOOLEAN'
-      elif isinstance(value, float):
-        return 'FLOAT'
-      elif isinstance(value, int):
-        return 'INTEGER'
-      else:
-        return 'STRING'
+    return [Schema._get_field_entry(name, value) for name, value in data.items()]
+
+  @staticmethod
+  def _from_list_record(data):
+    """
+    Infer a BigQuery table schema from a list of values.
+
+    Args:
+      data: The list of values.
+    Returns:
+      A list of dictionaries containing field 'name' and 'type' entries, suitable for use in a
+          BigQuery Tables resource schema.
+    """
+    return [Schema._get_field_entry('Column%d' % (i + 1), value) for i, value in enumerate(data)]
 
-    datum = data[0]
-    if isinstance(datum, dict):
-      return [{'name': key, 'type': _get_type(datum[key])} for key in datum.keys()]
+  @staticmethod
+  def _from_record(data):
+    """
+    Infer a BigQuery table schema from a list of fields or a dictionary. The typeof the elements
+    is used. For a list, the field names are simply 'Column1', 'Column2', etc.
+
+    Args:
+      data: The list of fields or dictionary.
+    Returns:
+      A list of dictionaries containing field 'name' and 'type' entries, suitable for use in a
+          BigQuery Tables resource schema.
+    """
+    if isinstance(data, dict):
+      return Schema._from_dict_record(data)
+    elif isinstance(data, list):
+      return Schema._from_list_record(data)
     else:
-      return [{'name': 'Column%d' % (i + 1), 'type': _get_type(datum[i])}
-              for i in range(0, len(datum))]
+      raise Exception('Cannot create a schema from record %s' % str(data))
 
   @staticmethod
-  def from_list(data):
+  def from_record(source):
     """
-    Infer a BigQuery table schema from a list. The list must be non-empty and be a list
-    of dictionaries (in which case the first item is used), or a list of lists. In the latter
-    case the type of the elements is used and the field names are simply 'Column1', 'Column2', etc.
+    Infer a BigQuery table schema from a list of fields or a dictionary. The type of the elements
```

Comment:
nit: Infer a BigQuery table schema from a single record (row) which can be a list of fields or a dictionary...

Just want to emphasize the difference between from_record and from_data